src/tree_data.h

```diff
@@ -854,6 +854,14 @@ struct lyd_node_term {
     };                                      /**< common part corresponding to ::lyd_node */
 
     struct lyd_value value;          /**< node's value representation */
+    struct lyd_node_term **leafref_nodes;   /**< list of the leafref data nodes [sized array](@ref sizedarrays)).
+                                                 By default it is empty. It is filled automatically by validation
+                                                 function of leafref nodes. It can also be manipulated manually by
+                                                 using [link api](@ref lyd_link_leafref_node),
+                                                 [unlink](@ref lyd_unlink_leafref_node). Freeing of the resources is
+                                                 automatic. */
+    struct lyd_node_term *target_node;      /**< pointer to leafref target data node, by default is NULL. The logic
+                                                 is the same as for [leafref_nodes](@ref leafref_nodes). */
```

Comment:
Shifting this to lyd_value doesn't make sense as it will still expand size of 2 pointers to all possible value, so somehow it will not help from memory perspective.
What is needed is the following:
- For all leafref data term nodes it should be stored the `target_node`
- For all data term nodes (including leafrefs) it is needed to store list of `leafref_nodes`
I was originally thinking to use `priv` for this purpose, but as I need to store both sections it would be needed to create some special structure for this + memory mgmt. operations for this. If it is better idea, then I rewrite it like this using such approach.

Reply:
Ok, not sure what memory magic you are referring to, but all the utest were successful including those using valgrind, so I think that memory magic still works. Anyway, if we really want to put this outside `lyd_node_term` then the only thing which can help is the proposed solution using hash_table. If I rewrite it using such approach, will it be acceptable from your PoV?

Reply:
Where would you put a pointer to this hash table? I am also thinking about another solution:

1. Add support for a new optional context callback that would be called for every validated data node in this context.
2. Write one such a public callback that would create a structure with all the leafref backlinks (and maybe also the target for a leafref) and set it to the `priv` pointer.

You said you have some other data in the `priv` pointer but then you could simply write your own callback that would first call this public one and then adjust the `priv` pointer in any way you need. Also, the second point is just so that the functionality is in *libyang* but I do not really care for that, up to you. What do you think?

Reply:
I have rewritten this using hash table stored within the context, I am not sure if this is better or not, up to you to judge. I personally don't like it as it seems quite strange solution. Anyway it works.

Yes I can make this extension using `priv` + callbacks on my side, but that rather complicates it my side, my project is not in C, but rather in python and I am using [libyang-python c](https://github.com/CESNET/libyang-python/), which doesn't really goes easily with additional C code. On top of that I would need callback not just for validation, but also for node deletion, unlink, value change & duplication.... so changing this like that seems to be even more complicated then having it directly under one C project...

Reply:
The solution seems fine to me.